src/ResourceManager/Common/Commands.ResourceManager.Common/Location/LocationCompleter.cs

```diff
@@ -0,0 +1,200 @@
+            {
+                if (resourceType != null)
+                {
+                    if (resourceTypeLocationDictionary.ContainsKey(resourceType))
+                        validResourceTypes.Add(resourceType);
+#if DEBUG
+                    else throw new Exception(String.Format(Resources.InvalidResourceType, resourceType));
+#endif
+                }
+            }
+
+            string[] distinctLocations;
+            if (validResourceTypes.Count > 0)
+            {
+                distinctLocations = resourceTypeLocationDictionary[validResourceTypes[0]].ToArray();
+                foreach (string resourceType in validResourceTypes)
+                {
+                    distinctLocations = distinctLocations.Intersect(resourceTypeLocationDictionary[resourceType].ToArray(), new LocationEqualityComparer()).ToArray();
+                }
+            }
+#if DEBUG
+            else throw new Exception(Resources.NoValidProviderFound);
+
+            if (distinctLocations.Length < 1) throw new Exception(Resources.NoValidLocationsFound);
+#endif
+            for (int i = 0; i < distinctLocations.Length; i++)
+            {
+                distinctLocations[i] = "\"" + distinctLocations[i] + "\"";
+            }
+            return distinctLocations;
+        }
+
+        /// <summary>
+        /// Create a dictionary mapping the ResourceTypes of all providers given to the valid locations available to that ResourceType
+        /// </summary>
+        /// <param name="allProviders"></param>
+        /// <returns></returns>
+        public static IDictionary<string, ICollection<string>> CreateLocationDictionary(List<Provider> allProviders)
+        {
+            IDictionary<string, ICollection<string>> resourceTypeLocationDictionary = new Dictionary<string, ICollection<string>>(StringComparer.OrdinalIgnoreCase);
+
+            foreach (var provider in allProviders)
+            {
+                foreach (var resourceType in provider.ResourceTypes)
+                {
+                    resourceTypeLocationDictionary.Add(String.Format("{0}/{1}", provider.NamespaceProperty, resourceType.ResourceType), resourceType.Locations);
+                }
+            }
+
+            return resourceTypeLocationDictionary;
+        }
+
+        private static ScriptBlock CreateScriptBlock(string[] resourceTypes)
+        {
+            string scriptResourceTypeList = "{" + String.Join(",", resourceTypes) + "}";
+            string script = "param($commandName, $parameterName, $wordToComplete, $commandAst, $fakeBoundParameter)\n" +
+                String.Format("$locations = [Microsoft.Azure.Commands.ResourceManager.Common.Location.LocationCompleterAttribute]::FindLocations({0})\n", scriptResourceTypeList) +
+                "$locations | Where-Object { $_ -Like \"`\"$wordToComplete*\" } | Sort-Object | ForEach-Object { [System.Management.Automation.CompletionResult]::new($_, $_, 'ParameterValue', $_) }";
```

Comment:
I am not sure aI follow what WordToComplete represents here, does this allow us to match locatiosn that match a particular string?

Reply:
This allows the user to type in part of the location (ex: west) and receive back only the locations that start with that phrase (ex: "West US", "West UK", ...)